### PR TITLE
Gh-3274: Simplified federation proof of concept

### PR DESCRIPTION
**Pull request by tb06904**
This is a basic proof of concept to demonstrate a simpler federated store with a rethink on how merging works and the required classes. 

This is **not** feature complete, the idea is this serves as a base for some additional required features. Once feature complete this may be able to replace the existing federated store to address its issues and technical debt.

Features present:
* Extendable merge/aggregator classes that support partially lazy element aggregation based on schemas.
* Ability to add graphs via the stores java methods.
* Operation handler for delegation to sub graphs (both for operations with and without output).
* Basic federated handler for merging schemas.
* Limited unit and integration testing.
* Ability to pass graph IDs and a toggle for element aggregation via operation options.

### CODE REVIEW

File: store-implementation/simple-federated-store/src/test/java/uk/gov/gchq/gaffer/federated/simple/merge/DefaultResultAccumulatorTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        Iterable<Entity> expected = () -> Arrays.asList(
+            new Entity("person", "1"),
+            new Entity("person", "1")).iterator();
+
+        FederatedResultAccumulator<Iterable<Entity>> accumulator = new DefaultResultAccumulator<>();
+        assertThat(accumulator.apply(iter1, iter2)).containsExactlyElementsOf(expected);
+    }
+
+    @Test
+    void shouldProvideElementAggregationBySchema() {
```
Comment (p29876) [inline store-implementation/simple-federated-store/src/test/java/uk/gov/gchq/gaffer/federated/simple/merge/DefaultResultAccumulatorTest.java:71]: Would be good to bulk out these tests to show how it doesn't merge 2 elements that are different and another where the graphs have multiple elements each that may or may not be merged but appreciate this is just the PoC

**Review by p29876 — DISMISSED** [review]
LGTM :fire: 
Only a couple nitpicks but this is just a PoC so can be left for now if you want

File: store-implementation/simple-federated-store/src/main/java/uk/gov/gchq/gaffer/federated/simple/FederatedStore.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+     *
+     * @throws IllegalArgumentException If graph not found.
+     */
+    public GraphSerialisable getGraph(final String graphId) {
+        for (final GraphSerialisable graph : graphs) {
+            if (graph.getGraphId().equals(graphId)) {
+                return graph;
+            }
+        }
+        throw new IllegalArgumentException("Graph for Graph ID: '" + graphId + "' is not available to this federated store");
```
[the remark this answers is not in the document]
Comment (tb06904) [reply store-implementation/simple-federated-store/src/main/java/uk/gov/gchq/gaffer/federated/simple/FederatedStore.java:90]: True but that would mean creating a stream and doing a `findFirst()` which would probably be more expensive and more code once the returned optional is handled.

File: store-implementation/simple-federated-store/src/test/java/uk/gov/gchq/gaffer/federated/simple/merge/DefaultResultAccumulatorTest.java (hunk 1 above)
Comment (tb06904) [reply store-implementation/simple-federated-store/src/test/java/uk/gov/gchq/gaffer/federated/simple/merge/DefaultResultAccumulatorTest.java:71]: Added an extra one but I agree these will need fleshing out.

**Review by p29876 — APPROVED** [review]

**Review by j69772 — DISMISSED** [review]

**Review by j69772 — APPROVED** [review]

**Review by wb36499 — APPROVED** [review]

**Pull request merged**